src/wire/dhcp.rs

```diff
@@ -0,0 +1,416 @@
+            9 => Ok(HardwareAddressType::Lanstar),
+            _ => Err(Error::Malformed)
+        }
+    }
+
+    pub fn hardware_address_len(&self) -> u8 {
+        self.buffer.as_ref()[field::HLEN]
+    }
+
+    pub fn transaction_id(&self) -> u32 {
+        let field = &self.buffer.as_ref()[field::XID];
+        NetworkEndian::read_u32(field)
+    }
+
+    pub fn client_hardware_address(&self) -> EthernetAddress {
+        let field = &self.buffer.as_ref()[field::CHADDR];
+        EthernetAddress::from_bytes(field)
+    }
+
+    pub fn hops(&self) -> u8 {
+        self.buffer.as_ref()[field::HOPS]
+    }
+
+    pub fn secs(&self) -> u16 {
+        let field = &self.buffer.as_ref()[field::SECS];
+        NetworkEndian::read_u16(field)
+    }
+
+    pub fn magin_number(&self) -> u32 {
+        let field = &self.buffer.as_ref()[field::MAGIC_NUMBER];
+        NetworkEndian::read_u32(field)
+    }
+
+    pub fn client_ip(&self) -> Ipv4Address {
+        let field = &self.buffer.as_ref()[field::CIADDR];
+        Ipv4Address::from_bytes(field)
+    }
+
+    pub fn your_ip(&self) -> Ipv4Address {
+        let field = &self.buffer.as_ref()[field::YIADDR];
+        Ipv4Address::from_bytes(field)
+    }
+
+    pub fn server_ip(&self) -> Ipv4Address {
+        let field = &self.buffer.as_ref()[field::SIADDR];
+        Ipv4Address::from_bytes(field)
+    }
+
+    pub fn relay_agent_ip(&self) -> Ipv4Address {
+        let field = &self.buffer.as_ref()[field::GIADDR];
+        Ipv4Address::from_bytes(field)
+    }
+
+    pub fn broadcast_flag(&self) -> Result<bool> {
+        let field = &self.buffer.as_ref()[field::FLAGS];
+        match NetworkEndian::read_u16(field) {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(Error::Malformed),
+        }
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> Packet<&'a T> {
+    /// Return a pointer to the options.
+    #[inline]
+    pub fn options(&self) -> Result<&'a [u8]> {
+        let data = self.buffer.as_ref();
+        data.get(field::OPTIONS).ok_or(Error::Malformed)
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    pub fn set_opcode(&mut self, value: u8) {
+        assert!(value == 1 || value == 2);
+        let data = self.buffer.as_mut();
+        data[field::OP] = value;
+    }
+}
+
+/// The possible opcodes/message types of a DHCP packet.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum MessageType {
+    Discover,
+    Offer,
+    Request,
+    Decline,
+    Ack,
+    Nak,
+    Release,
+    Inform,
+}
+
+impl MessageType {
+    fn opcode(&self) -> u8 {
+        match *self {
+            MessageType::Discover | MessageType::Inform | MessageType::Request |
+                MessageType::Decline | MessageType::Release => 1,
+            MessageType::Offer | MessageType::Ack | MessageType::Nak => 2,
+        }
+    }
+}
+
+/// The possible hardware address types of a DHCP packet.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum HardwareAddressType {
```

Comment:
You should take a look at how other enums are defined. There's a macro that writes conversion functions for you:
```
enum_with_unknown! {
    /// ARP operation type.
    pub enum Operation(u16) {
        Request = 1,
        Reply = 2
    }
}
```

Reply:
Looks good otherwise, though.

Reply:
That's useful, thanks!